### ISSUE TITLE
Transaction hash is not available to the user on contract creation

### ISSUE THREAD

**Issue by tgerring** [ethereum/go-ethereum#1306]
Creating a contract (i.e. transaction with "data" but without "to") returns the contract address to the user, making it difficult to tell if/when the transaction is mined. The adage is, "just check with getCode", however, it would be far better to report the transaction pool hash so that the transaction itself can be checked for inclusion in a block. This is how standard value transactions are tracked, for example by verifying on a block explorer.

Instead, I propose that the expected contract address is stored with the transaction in local storage. This benefits the user in multiple ways because it creates an explicit link between a transaction and resulting contract. This missing link is obvious when viewing a block explorer, for example, http://etherchain.org/account/0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b list associated transaction `0x4ad149e31b6f4e0f60b4bbfe2786201fca8004589bd7b28dd859e8f1649911b8` with no indication of the resulting contract, `0x66b36f8b34c13246dbe8651ee4cffc4a12cffecf`.

If the contract address were stored on a field of the transaction, programatic access is much simpler and will be a boon to developers. Currently, the user must capture the intended contract address on transaction creation immediately and if not captured, it is completely unavailable to reference at a later time unless they, i.e. view it in the pending transaction list before entering the network transaction pool.

Originally, I considered storing this link in the "to" field. This presents a problem, however, if the main distinguisher between types of transactions is the lack of "to" field. For example, in this structure, sending data to an existing contract and sending a transaction that results in a contract would be indistinguishable if "to" were used for multiple purposes. One solution to this is to store the contract address in a separate field or use a flag to indicate if the "to" address is the result of a contract creation.

I'm interested in hearing feedback on this, but ultimately, I'm of the conclusion that the transaction hash of an attempted contract creation should not be hidden from the user.

**Comment by joeykrug** [comment]
Hmmm it shows the hash when I make an rpc call to create a contract:

``` I0620 22:23:58.820025   13651 xeth.go:973] Tx(7f6e8ac415b5908ad45d5f50660e87d12e90f3b9875c08f4e38b277bfd354380) created: fbcdc346cb7da9f56c63b127b45fedf8b23278a0```

**Comment by tgerring** [comment]
@joeykrug This is only printed in the log and not presented to the user in RPC or Console.

**Comment by frozeman** [comment]
The actual solution would be to always return the transaction hash when transacting and storing the contract address (if one was created) in the transaction receipt (which you can get using the transaction hash after its mined). https://github.com/ethereum/go-ethereum/issues/1042

**Comment by obscuren** [comment]
see https://github.com/ethereum/go-ethereum/issues/1042

**Comment by tgerring** [comment]
Is there a related issue to update the console output of the transaction? It seems that the contract address is still being displayed rather than transaction hash

**Comment by tgerring** [comment]
Included the following change at dd521ece3ff6163c891d2f945903533079413903 for PR #1397:

```
diff --git a/xeth/xeth.go b/xeth/xeth.go
index cbc8dbb..1e87b73 100644
--- a/xeth/xeth.go
+++ b/xeth/xeth.go
@@ -966,7 +966,6 @@ func (self *XEth) Transact(fromStr, toStr, nonceStr, valueStr, gasStr, gasPriceS
        if contractCreation {
                addr := crypto.CreateAddress(from, nonce)
                glog.V(logger.Info).Infof("Tx(%x) created: %x\n", tx.Hash(), addr)
-               return addr.Hex(), nil
        } else {
                glog.V(logger.Info).Infof("Tx(%x) to: %x\n", tx.Hash(), tx.To())
        }
```

**Comment by obscuren** [comment]
:+1:

**Comment by obscuren** [comment]
This has been fixed in #1397
